### ISSUE TITLE
Migrate to Vue 2.x

### ISSUE THREAD

**Issue by cpb8010** [klembot/twinejs#461]
I was trying to include some spiffy new components and found that the project is still on vue 1.0.18 (and the old vuex)

If I started this work, is there any interest in getting this merged?
Is someone already doing this? Has someone tried and failed?

I'm starting work on this on my own fork and wanted to see if there was any traction here.

**Comment by klembot** [comment]
I've looked into it; the main obstacle I see is that Vue 2.x removed $broadcast, $dispatch, and $emit which are used. I'm open to a PR but would want to take a careful eye to make sure no regressions are introduced. Ideally, this migration would include adding tests on the Vue components, but I also realize that adds a lot of work.

**Comment by cpb8010** [comment]
Yes, looking into it those signal changes and the HTML interpolation
removal are the most disruptive.

**Comment by wraybowling** [comment]
I dove into twinejs for the first time today, and immediately noticed the same thing. @cpb8010 I'll happily assist with the effort.

**Comment by cpb8010** [comment]
Hey @wraybowling 

I was able to get through most of the migration helper, but I think the start-up process for Vue 2 is different so the mixins/plugins don't load quite the same.
I have the changes here:
https://github.com/cpb8010/twinejs/pull/1 / https://github.com/cpb8010/twinejs/tree/tx-pass3

If you want to pull it down to take a pass at it.
I'll be muddling around with it as I dig through the docs.

**Comment by cpb8010** [comment]
I figured part of it out. Vue 2 doesn't support require syntax the same way and the migration helper missed some of the router and app init changes. I'll post again when I have that sorted.

**Comment by cpb8010** [comment]
So I figured that part out, but since $appendTo was removed $mountTo is broken.

8 tests are failing. 4 with CodeMirror and 4 with $mountTo.

https://github.com/cpb8010/twinejs/pull/2

I think it's close, but I haven't gotten far enough to actually load a story in the application

**Comment by cpb8010** [comment]
Turns out the failing tests were because of the same init issues where it pulled in the version without the template compiler. I haven't figure out enough webpack to make this work without changing the test source.


I took the $appendTo source from Vue 1 to fix $mountTo, but I'm still not sure it works because it doesn't show any default stories

**Comment by wraybowling** [comment]
I wasn't able to get the repo to build.. it may be a little while before I can contribute

**Comment by cpb8010** [comment]
Hey @wraybowling I had some old branches pushed that wouldn't build, the most recent one is called fix-startup. Let me know if you have specific errors I can help with.

**Comment by wraybowling** [comment]
@cpb8010 Okay. I got my setup working. Now to make sense of what you're doing... I've been using the .vue file structure; not as used to seeing it all split up like this.

**Comment by cpb8010** [comment]
Nice!

The webpack setup doesn't include the vue compiler step, but it could!
I was trying to get it otherwise working before modernizing it.

**Comment by wraybowling** [comment]
@cpb8010 Where can I learn more about the webpack setup?
Yeah, it's fun using es2017 and .vue files; it's what lured me in to trying Vue. I think you're wise to upgrade one little piece at a time if you understand how to do so.

**Comment by cpb8010** [comment]
I've been comparing the webpack config in this project to a fresh config
created with vue-cli. This project is on an older version of web-pack as
well, so it's not one-to-one, but it's close.

The biggest part of the webpack in this project that's custom is for the
i18n stuff and NW.js, which I haven't even gotten close to yet.

**Comment by klembot** [comment]
Just popping in to say that while I could be talked into using .vue files, it would take some convincing. I understand that that is how you're supposed to do it in Vue 2.x, but I am hesitant because

- Although it is often possible in editors to switch syntax modes inside a single file, I've always found it to be a pain to set up... though I have not tried this in quite a while
- I worry that we would end up with thousand-line monstrosities for some of the more intricate components

I'm happy to chat more about the migration process and any other decisions you may need to make-- I would rather stay in regular contact as you guys work on this rather than having a big debate at the very end. Keep on truckin 🚚

**Comment by klembot** [comment]
p.s. on my todo list has been migrating from NW.js to Electron. Just saying in case you find yourself banging your head against a wall with NW.js.

**Comment by cpb8010** [comment]
Thanks!

I did briefly look into something like https://simulatedgreg.gitbooks.io/electron-vue/content/en/
so I'll cross that bridge when I come to it! 

The farthest I've gotten was just getting a story loaded and it not showing any of the elements on the grid.

**Comment by greyelf** [comment]
> migrating from NW.js to Electron

@klembot 
Could you explain your reasons for considering this migration?
(I'm not questioning it as I have little experience with either, but I'm looking to learn one of them in greater detail and knowing your reasons may help me choose.)

**Comment by klembot** [comment]
At a high level... Electron has more momentum behind it right now thanks to big projects like VS Code and Slack using it, and accordingly working with it has a lot less friction. The other thing I have liked more about it is that it enforces a separation between the "web" part and the part of the app that knows about the filesystem, etc. The Twine code right now patches behaviors inside the web app, which worked OK at first but is messy to debug and/or edit.

**Comment by cpb8010** [comment]
### More updates

After some debugging I found that it's difficult to emulate the existing modal templates' _thenable_ flow without broadcast and dispatch. The modal-dialog to prompt relationship in particular is what is tripping me up. The other part is that the migration helper didn't catch all of the old dispatches and event sections, so things were more broken then they should have been.

If I get more stuck I might end up refactoring to streamline and simplify this flow assuming nothing external is dependent on it.

**Comment by klembot** [comment]
Just curious, would you be able to chat about the state of this on the Discord sometime? I'm on the Eastern time zone.

**Comment by cpb8010** [comment]
Yes, I'll PM you tomorrow when I'm free.

After more fiddling I found that the transitions might be involved and
those were only partially migrated as well.

**Comment by klembot** [comment]
What's your username on Discord? I have a couple friend invites from strangers and I'm not sure which might be you.

**Comment by cpb8010** [comment]
crash_n_burn

**Comment by cpb8010** [comment]
Another update on this, I was able to get the modal prompt working by avoiding $mountTo/thenable.
There are another 15 uses of $mountTo which might also be subtly broken, but the workaround I did was to mount and hide the component on the virtual DOM, then trigger showing it with the data it needs over the global event bus. This didn't take long, but it's not the cleanest solution I'd wish worked.

As for the next problem, I don't know if this is what is breaking the story-edit-view. There are lots of errors (including a stack overflow coming from a Vuex action) that I'll need to figure out before making more progress.

https://github.com/cpb8010/twinejs/tree/change-prompt Is my latest branch.

If you made any progress on your experimental work, let me know and I'd also be glad to help there 🔨

**Comment by cpb8010** [comment]
So my PR is here: https://github.com/klembot/twinejs/pull/501

I changed only what I came across, so no big changes like single file components or electron.
Most of it was simple syntax changes, and the functional changes were small except for the $mountTo/thenable stuff mentioned above, which I've completely removed for static components in the parent template. I got really stuck because I was working in Edge and thought that I had broken the play button. Turns out Play with Edge doesn't work in the current twine either! Opera worked fine, so I assume Chrome works as well.

I don't expect 18k lines to be reviewed quickly or accepted easily, so I expect I'll have to keep the branch up to date as other smaller work goes in. Obviously if you have design questions or other requests I'll respond here, the PR, or Discord.

🦃
